### DISCUSSION TITLE
Persistence with custom properties

### DISCUSSION THREAD

**Discussion by tlf30** [andrewauclair/ModernDocking#111] [General]
Hello,
With the old persistence system, I was saving custom properties that were not directly tied to fields in the code.
Is there a way to accomplish this with the new system without creating a field for each property?

Also, I am running into an unsupported property type exception, I think it is on a boolean type. Could you please add booleans and make the exception more descriptive? Currently the exception does not have any information to identify what property is the issue. 

Thanks,
Trevor

**Comment by andrewauclair** [comment]
Maybe I could add support for Map<String, String> somehow to the new system. Best of both worlds. 

I did all the basic types and somehow forgot boolean? That's a pain. I'll get that added.

**Comment by tlf30** [comment]
Thanks,
I am thinking about the Map<String, String>. The old setup was a little clunky. Do you think you could provide some type of static access to the persistence that gets the same properties as the annotations and allows accessing existing annotated properties, and adding new ones? Perhaps it does not need to be static to implement it, but I have not done a deep dive to see how the new system works.

**Reply by andrewauclair** [reply]
Let me see if I understand what you're asking. Some way to basically write your own read/write handling for property annotations?

Should be doable. Maybe using reflection to find static functions with certain signatures that read and write in the type? This would have to be for custom user defined types only. But that should still make it possible to implement primitives I may have missed by just wrapping them in a class and writing the read/write.

**Reply by tlf30** [reply]
I was thinking something like `AppState.put(dockable, "property name", value)`

And an associated getter. This would allow for properties to be stored that do not have an associated annotated field. Putting a property does not necessarily need to write to an annotated field if one exists for the property, but that would be a cool feature.

**Reply by andrewauclair** [reply]
I like that.

How about:
```java
public static String getProperty(Dockable dockable, String propertyName);
public static void setProperty(Dockable dockable, String propertyName, String value);
```

These properties would get saved together with the rest of the properties through `DockableProperties`. They will also be loaded together, any property that is not associated with a field in the `Dockable` will be set in `AppState` before `updateProperties` is called.

Maybe there's a way to keep the type for these ones as well. I just can't think of it right now.

**Reply by tlf30** [reply]
That would work great. Preserving the type would be very nice, but for now strings would be fine.

**Comment by i-make-robots** [comment]
https://github.com/andrewauclair/ModernDocking/wiki/Persistence

Please, how do i use persistence?  I don't see it in the wiki, the documentation, or the single-app examples.

**Reply by andrewauclair** [reply]
I have this planned for the new documentation. But the essentials can be found in the basic example. Snippet below:

```java
SwingUtilities.invokeLater(() -> {

    setVisible(true);

    // now that the main frame is set up with the defaults, we can restore the layout
    AppState.setPersistFile(layoutFile);

    try {
        AppState.restore();
    } catch (DockingLayoutException e) {
        // something happened trying to load the layout file, record it here
        e.printStackTrace();
    }

    AppState.setAutoPersist(true);
});
```

Auto persistence defaults to off and has to be turned on by the application. Persistence works off of a 500ms timer which is fired any time `AppState.persist` is called. Either manually by the application or automatically by the docking framework when dockables are moved, windows resized, splits resized, etc.

Layouts can also be saved and loaded from your own files with a combination of `DockingState.getApplicationLayout()` and `LayoutPersistence.saveLayoutToFile` for saving and `LayoutPersistence.loadApplicationLayoutFromFile` and `DockingState.restoreApplicationLayout` for loading.

I will link the new documentation here when it is ready.

**Comment by i-make-robots** [comment]
should I be concerned about the very angry looking messages in my log file?

![image](https://github.com/andrewauclair/ModernDocking/assets/1464454/58faa0b9-343b-4d7e-b42e-eaff1d34a68e)

**Reply by andrewauclair** [reply]
No, it's just an info. Not sure why an info gets colored red. That's a little weird.

I probably should set that to debug instead of info. Mostly a sanity check to myself that the persist is always working.

**Comment by i-make-robots** [comment]
Finally back on this quesiton.... so.... say I make class `Preview` and class `Tree`, both of which `extends JPanel implements Dockable` and both `Docking.registerDockable(this);` in their constructors.  When `AppState.restore()` is called, how do I make sure the correct classes are reinstanced?

**Reply by tlf30** [reply]
Not sure I fully understand the question. You create the instance when you register the dockable. You need to register them every time the app runs prior to restoring the layout. 

The layout looks at the persistence id of the dockable to know which one is associated with the persisted data. You need to have a unique id for every dockable instance.

**Reply by i-make-robots** [reply]
My program has plugins which may create dockable tabs.  At program start none of them creates a window, apart from the required root.  I need a way to look at the dockables that are restored, find the properties, and then put the right plugin jpanel into each of those dockables.  I believe I'll use the properties to record the plugin and any extra data (like which asset is being edited by this plugin).

**Reply by tlf30** [reply]
I am also using a plugin system. I simply show the UI after plugins are initialized. I use a loading bar jframe to display plugin iniy progress

**Reply by i-make-robots** [reply]
So.... you create one .... permanent jpanel.... for each plugin?  Part of my goal is to be able to dispose and recreate panels so i can see gui changes in real time.

**Reply by tlf30** [reply]
No, each plugin registers as many as it wants. You can dynamically register additional panels at runtime, but if you want to restore properties to them, they need to exist when you restore the app state. You can always create and destroy panels with the same persistent id.

**Reply by i-make-robots** [reply]
So.... before I restore state I need to know how many windows will be restored and create them with the matching persistent ID to make sure restore works ok?  That sounds to me a bit chicken and egg.

**Reply by andrewauclair** [reply]
Modern Docking can recreate the dockables on the fly when reloading the application. You either need a constructor with `String tabText, String title, String persistentID` (this is the old way) or a constructor with `DynamicDockableParameters parameters` (new way). You then do the normal steps to build your panel and register the dockable. When the properties are loaded, Modern Docking will call the dockable method `updateProperties`, which you can override. When it is called, any `@DockingProperty` members will have been loaded. See https://github.com/andrewauclair/ModernDocking/blob/2b50d863890049370bbb29e4b02159418f89f3dc/demo-single-app/src/basic/SimplePanel.java for an example of a class configured to be dynamic.

**Reply by i-make-robots** [reply]
Does MD remember which class type is extending Dockable and rely on that specific public constructor existing?
I'm concerned that if MD instantiates the class then the instance won't be registered with my factory.

**Reply by andrewauclair** [reply]
Yes. It persists the class name in the layout. Then created the class instance with the constructor call. Otherwise the restore fails and that dockable is not restored.

**Reply by tlf30** [reply]
> Modern Docking can recreate the dockables on the fly when reloading the application. You either need a constructor with `String tabText, String title, String persistentID` (this is the old way) or a constructor with `DynamicDockableParameters parameters` (new way). You then do the normal steps to build your panel and register the dockable. When the properties are loaded, Modern Docking will call the dockable method `updateProperties`, which you can override. When it is called, any `@DockingProperty` members will have been loaded. See https://github.com/andrewauclair/ModernDocking/blob/2b50d863890049370bbb29e4b02159418f89f3dc/demo-single-app/src/basic/SimplePanel.java for an example of a class configured to be dynamic.

I did not know this. Do you think you could add a way to handle when a custom constructor is required? My current solution is to store what tabs were open and register them prior to restoring the app state. But if there was a way to set a custom way to create the dockables, like a callable for requesting to create a dockable and let custom logic handle the creation?

**Reply by i-make-robots** [reply]
Yeah, that's what I was looking for, too.  In the current paradigm if I change a class name (unlikely) any user from the previous version will get an error and then a UI with bits missing.  If I can control it myself (calling my panel factory) then I can version control for name changes.

**Reply by andrewauclair** [reply]
> I did not know this. Do you think you could add a way to handle when a custom constructor is required? My current solution is to store what tabs were open and register them prior to restoring the app state. But if there was a way to set a custom way to create the dockables, like a callable for requesting to create a dockable and let custom logic handle the creation?

Sounds possible to me. It could be a callback to overrides the current behavior. Modern Docking can pass all the information it knows about the panel from the layout: tab text, title text, persistent ID, and class name.

**Reply by tlf30** [reply]
That would work well. I encode info in my persistence id so I can recreate the panel in my current way of handling it. Having a callback to create custom panel to override the current behavior would be perfect. 
Really all I need is the persistence id, but I imagine most people will want the other info such as the class name.
